src/main/java/backend/Logic.java

```diff
@@ -93,21 +95,54 @@ public void refresh() {
                 .exceptionally(Futures::log);
     }
 
+    /**
+     * Opens the repoId if it isn't already open, else simply refreshes the UI
+     * After opening the repo, it will trigger a PrimaryRepoOpenedEvent
+     *
+     * @param repoId
+     * @return
+     */
     public CompletableFuture<Boolean> openPrimaryRepository(String repoId) {
-        return openRepository(repoId, true);
+        return openRepository(repoId, true, Optional.empty());
     }
 
-    public CompletableFuture<Boolean> openRepositoryFromFilter(String repoId) {
-        return openRepository(repoId, false);
+    /**
+     * Opens the repoId if it isn't already open, else simply refreshes the UI
+     * After opening the repo, it will trigger a FilterOpeningEvent with panel as argument
+     *
+     * @param repoId
+     * @return
+     */
+    public CompletableFuture<Boolean> openRepositoryFromFilter(String repoId, FilterPanel panel) {
+        return openRepository(repoId, false, Optional.of(panel));
     }
 
-    private CompletableFuture<Boolean> openRepository(String repoId, boolean isPrimaryRepository) {
+    /**
+     * Opens the repoId if it isn't already open, else simply refreshes the UI
+     *
+     * After opening the repo, it will trigger a PrimaryRepoOpenedEvent or AppliedFilterEvent depending on the
+     * caller origin determined by isPrimaryRepository
+     *
+     * @param repoId
+     * @param isPrimaryRepository false if it was due to repo being specified in a panel's filter
+     * @param panel should be Optional.empty() if isPrimaryRepository is true, else it should contain
+     *              the panel whose filter specified the repo
```

Comment:
`isPrimaryRepository` and `panel` parameters seem to be very closely coupled i.e. the value of one parameter requires another value for the other. Is it better to use just 1 parameter for this? Then you can declare a variable `isPrimaryRepository` inside the method based on `panel`'s value to keep things readable.